pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -249,6 +255,74 @@ func (rc *reconciler) mountAttachedVolumes(volumeToMount cache.VolumeToMount, po
 	}
 }
 
+// processReconstructedVolumes checks volumes which were skipped during the reconstruction
+// process because it was assumed that since these volumes were present in DSOW they would get
+// mounted correctly and make it into ASOW.
+// But if mount operation fails for some reason then we still need to mark the volume as uncertain
+// and wait for the next reconciliation loop to deal with it.
+func (rc *reconciler) processReconstructedVolumes() {
+	if rc.kubeClient != nil {
+		rc.updateDevicePath(rc.skippedDuringReconstruction)
+	}
+	for volumeName, glblVolumeInfo := range rc.skippedDuringReconstruction {
+		// check if volume is marked as attached to the node
+		// for now lets only process volumes which are at least known as attached to the node
+		// this should help with most volume types (including secret, configmap etc)
+		if !rc.actualStateOfWorld.VolumeExists(volumeName) {
+			klog.V(4).InfoS("Volume is not marked as attached to the node. Skipping processing of the volume", "volumeName", volumeName)
+			delete(rc.skippedDuringReconstruction, volumeName)
+			continue
+		}
+		uncertainVolumeCount := 0
+
+		for podName, volume := range glblVolumeInfo.podVolumes {
+			volumeNotMounted := rc.actualStateOfWorld.PodRemovedFromVolume(podName, volume.volumeName)
+			// if volume is not mounted then lets mark volume mounted in uncertain state in ASOW
+			if volumeNotMounted {
+				markVolumeOpts := operationexecutor.MarkVolumeOpts{
+					PodName:             volume.podName,
+					PodUID:              types.UID(volume.podName),
+					VolumeName:          volume.volumeName,
+					Mounter:             volume.mounter,
+					BlockVolumeMapper:   volume.blockVolumeMapper,
+					OuterVolumeSpecName: volume.outerVolumeSpecName,
+					VolumeGidVolume:     volume.volumeGidValue,
+					VolumeSpec:          volume.volumeSpec,
+					VolumeMountState:    operationexecutor.VolumeMountUncertain,
+				}
+				err := rc.actualStateOfWorld.AddVolumeViaReconstruction(markVolumeOpts)
```

Comment:
there might be a race? when you check actual state previously, podvolume is not in actual state, but when you add volume next, it is possible that volume is added into the state.

Reply:
Yes there is a small window where `AddVolumeViaReconstruction` could race with `MarkVolumeAsMounted` or something similar operation. I have thought about it and if `AddVolumeViaReconstruction` is last function to run then mount operation will start again (because we marked something as uncertain which was previously marked as mounted).  I think it should be fine since mount operation is anyways idempotent.

But - it is also possible to fix this race with somewhat heavy handed approach to locking. If we held lock for `PodRemovedFromVolume` and `AddVolumeViaReconstruction` both functions with ever releasing it(say we made a single function that does `CheckAndMarkAsUncertain`), then I believe race is eliminated because either this function will wait for lock or other function will wait for lock. 

I personally do not mind leaving the code as it is, but I am happy to change if you think it is worth fixing for race condition. what do you think?

Reply:
Will another race occur?
1. desired state still has the volume since actual state does not has it yet, so it is in the process of mounting
2. volume is added into the actual state through the logic added here.
3. desired state populator checks the actual state has the volume already, and delete volume from desired state
4. the pending mount operation from 1 failed, and mark volume as unmounted.

in this case, pod will not stuck in terminating any more, but mount is not cleaned up?

Reply:
> the pending mount operation from 1 failed, and mark volume as unmounted.

If volume was added to ASOW via reconstruction then a failed mount won't mark volume as unmounted. It will still remain in uncertain state and volume must be unmounted for pod to be deleted.

Reply:
oh, right, I forgot you changed the logic there too.

Reply:
I replaced above function with `CheckAndMarkVolumeAsUncertainViaReconstruction`.